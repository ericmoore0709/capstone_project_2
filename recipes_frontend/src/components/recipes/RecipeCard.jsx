import { Button, Card, CardBody, CardImg, CardText, CardTitle, DropdownItem, DropdownMenu, DropdownToggle, UncontrolledDropdown } from 'reactstrap';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import defaultImage from '../../assets/defaultImage.jpg';

const RecipeCard = ({ recipe, signedInUser, deleteRecipe, shelfOptions, addRecipeToShelf, handleRemoveRecipeFromShelf }) => {

    const handleAddRecipeToShelf = async (shelfId) => {
        addRecipeToShelf(shelfId, recipe.id);
    }

    const deletionConfirmed = (msg) => {
        return window.confirm(msg);
    }

    const handleOnDelete = (id) => {
        if (deletionConfirmed('Are you sure you want to delete this recipe?')) {
            deleteRecipe(id);
        }
    }

    return (
        <div className='p-2 d-flex'>
            <Card style={{ minWidth: '200px', minHeight: '250px', width: '23vw' }}>

                {/* At-a-glance recipe info */}
                <Link to={`/recipes/${recipe.id}`}><CardImg src={recipe.image || defaultImage} alt={recipe.title} style={{ maxHeight: '150px', objectFit: 'cover' }} /></Link>
                {shelfOptions &&
                    <UncontrolledDropdown className='position-absolute top-0 end-0'>
                        <DropdownToggle color='primary'>
                            +
                        </DropdownToggle>
                        <DropdownMenu end>
                            <div className='ms-2'>Add to shelf...</div>
                            <DropdownItem divider />
                            {shelfOptions.map((shelf) =>
                                <DropdownItem key={shelf.id} className='text-center' onClick={() => handleAddRecipeToShelf(shelf.id)}>{shelf.label}</DropdownItem>
                            )}
                        </DropdownMenu>
                    </UncontrolledDropdown>
                }

                {handleRemoveRecipeFromShelf &&
                    <Button onClick={() => handleRemoveRecipeFromShelf(recipe.id)} color='danger' className='position-absolute top-0 end-0'>-</Button>
                }

                <CardBody className='d-flex flex-column'>
                    <CardTitle tag='h5' className='text-center'><Link to={`/recipes/${recipe.id}`}>{recipe.title}</Link></CardTitle>
                    <CardText className='text-truncate' style={{ maxHeight: '3em', overflow: 'hidden' }}>{recipe.description}</CardText>
                </CardBody>

                {/* Meta Info*/}
                {recipe.author && <>
                    <small>
                        <div className='d-flex'>
                            <div className='w-25 text-start mx-1'>
                                <div><strong>Uploaded</strong></div>
                                <div>{new Date(recipe.uploaded_at).toLocaleDateString()}</div>
                            </div>
                            <div className='w-50 text-center mx-auto'>
                                <div><strong>Author</strong></div>
                                <div><Link to={`/profiles/${recipe.author.id}`}>{`${recipe.author.firstName} ${recipe.author.lastName}`}</Link></div>
                            </div>
                            <div className='w-25 text-end mx-1'>
                                <div><strong>Last Updated</strong></div>
                                <div>{new Date(recipe.last_updated_at).toLocaleDateString()}</div>
                            </div>
                        </div>
                    </small>

                    {/* Author controls */}
                    {(recipe.author_id === signedInUser.id) && <div className='text-center'>
                        <Link to={`/recipes/${recipe.id}/edit`} className='w-50 btn btn-secondary mb-1'>Edit</Link>
                        <Button color='danger' className='w-50 mb-1' onClick={() => handleOnDelete(recipe.id)}>Delete</Button>
                    </div>}
                </>}
            </Card>
        </div>
    );
}

RecipeCard.propTypes = {
    recipe: PropTypes.object.isRequired,
    signedInUser: PropTypes.object,
    deleteRecipe: PropTypes.func,
    shelfOptions: PropTypes.array,
    addRecipeToShelf: PropTypes.func,
    handleRemoveRecipeFromShelf: PropTypes.func
};

export default RecipeCard;
