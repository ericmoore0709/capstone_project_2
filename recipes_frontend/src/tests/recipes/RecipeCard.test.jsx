import { fireEvent, render, screen } from "@testing-library/react";
import { afterEach, describe, expect, it, vi } from "vitest";
import { MemoryRouter } from "react-router-dom"; // Required for Link components
import RecipeCard from "../../components/recipes/RecipeCard";
import defaultImage from '../../assets/defaultImage.jpg';

describe('RecipeCard', () => {
    const mockRecipe = {
        id: 1,
        title: "Recipe 1",
        description: "Tasty!",
        image: null,
        author_id: 1
    };

    const mockSignedInUser = { id: 1 };
    const mockDelete = vi.fn();

    const renderComponent = (recipe = mockRecipe, signedInUser = mockSignedInUser) => {
        return render(
            <MemoryRouter>
                <RecipeCard recipe={recipe} signedInUser={signedInUser} deleteRecipe={mockDelete} />
            </MemoryRouter>
        );
    };

    afterEach(() => {
        vi.restoreAllMocks();
    });

    it('renders without crashing', () => {
        renderComponent();
    });

    it('takes a snapshot', () => {
        const { asFragment } = renderComponent();
        expect(asFragment()).toMatchSnapshot();
    });

    it('correctly displays title and description', () => {
        renderComponent();
        expect(screen.getByText(mockRecipe.title)).toBeInTheDocument();
        expect(screen.getByText(mockRecipe.description)).toBeInTheDocument();
    });

    it('displays the default image if no image is provided', () => {
        renderComponent();
        const img = screen.getByRole('img');
        expect(img).toHaveAttribute('src', defaultImage);
        expect(img).toHaveAttribute('alt', mockRecipe.title);
    });

    it('displays the recipe image if provided', () => {
        const customRecipe = { ...mockRecipe, image: "https://via.placeholder.com/150" };
        renderComponent(customRecipe);
        const img = screen.getByRole('img');
        expect(img).toHaveAttribute('src', customRecipe.image);
    });

    it('correctly displays "See more!" button', () => {
        renderComponent();
        const seeMoreLink = screen.getByRole('link', { name: /see more!/i });
        expect(seeMoreLink).toBeInTheDocument();
        expect(seeMoreLink).toHaveAttribute('href', `/recipes/${mockRecipe.id}`);
    });

    it('conditionally displays "Edit" button for author', () => {
        renderComponent();
        const editLink = screen.getByRole('link', { name: /edit/i });
        expect(editLink).toBeInTheDocument();
        expect(editLink).toHaveAttribute('href', `/recipes/${mockRecipe.id}/edit`);
    });

    it('does not display "Edit" button for non-author', () => {
        const nonAuthorUser = { id: 2 }; // Different user ID
        renderComponent(mockRecipe, nonAuthorUser);
        const editButton = screen.queryByRole('link', { name: /edit/i });
        expect(editButton).not.toBeInTheDocument();
    });

    it('conditionally displays "Delete" button for author', () => {
        renderComponent();
        const deleteButton = screen.getByRole('button', { name: /delete/i });
        expect(deleteButton).toBeInTheDocument();
    });

    it('does not display "Delete" button for non-author', () => {
        const nonAuthorUser = { id: 2 }; // Different user ID
        renderComponent(mockRecipe, nonAuthorUser);
        const deleteButton = screen.queryByRole('button', { name: /delete/i });
        expect(deleteButton).not.toBeInTheDocument();
    });

    it('handles delete button click with confirm', () => {
        vi.spyOn(window, 'confirm').mockImplementation(() => true);
        renderComponent();
        const deleteButton = screen.getByRole('button', { name: /delete/i });
        fireEvent.click(deleteButton);
        expect(mockDelete).toHaveBeenCalled();
    });

    it('handles delete button click with reject', () => {
        vi.spyOn(window, 'confirm').mockImplementation(() => false);
        renderComponent();
        const deleteButton = screen.getByRole('button', { name: /delete/i });
        fireEvent.click(deleteButton);
        expect(mockDelete).not.toHaveBeenCalled();
    });
});
