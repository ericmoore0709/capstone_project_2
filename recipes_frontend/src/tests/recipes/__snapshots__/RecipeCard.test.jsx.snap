// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RecipeCard > takes a snapshot 1`] = `
<DocumentFragment>
  <div
    class="p-2 d-flex"
  >
    <div
      class="card"
      style="min-width: 200px; min-height: 250px; width: 23vw;"
    >
      <a
        href="/recipes/1"
      >
        <img
          alt="Recipe 1"
          class="card-img"
          src="/src/assets/defaultImage.jpg"
          style="max-height: 150px; object-fit: cover;"
        />
      </a>
      <div
        class="d-flex flex-column card-body"
      >
        <h5
          class="text-center card-title"
        >
          <a
            href="/recipes/1"
          >
            Recipe 1
          </a>
        </h5>
        <p
          class="text-truncate card-text"
          style="max-height: 3em; overflow: hidden;"
        >
          Tasty!
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
